docs/guides/flax_fundamentals/rng_guide.md

```diff
@@ -0,0 +1,650 @@
+
+x = jnp.ones((1, 3))
+y = jnp.ones((1, 2))
+
+module = Model()
+variables = module.init(init_rngs, x, train=False)
+```
+
+```{code-cell}
+:outputId: a99113f7-aca7-4914-f592-5af674673a1a
+
+def update(variables, rng):
+  # we don't need to provide a 'params' or 'other' rng, as only 'dropout' rng will be used during training
+  # split the rng to get a dropout_rng to be used for this training iteration,
+  # and to get another rng key to be used for the next training iteration
+  dropout_rng, next_rng = jax.random.split(rng)
+  def loss(params):
+    out = module.apply({'params': params, 'other_collection': variables['other_collection']}, x, train=True, rngs={'dropout': dropout_rng})
+    return jnp.mean((y - out)**2)
+  grads = jax.grad(loss)(variables['params'])
+  params = jax.tree_map(lambda p, g: p - 1e-3 * g, variables['params'], grads)
+  return {'params': params, 'other_collection': variables['other_collection']}, next_rng
+
+for _ in range(10):
+  variables, train_rng = update(variables, train_rng)
+  out = module.apply(variables, x, train=False)
+  print(jnp.mean((y - out)**2))
+```
+
+### 🔪 Sharp edge 🔪 - unintentionally generating the same values
+
+There is an edge case where the same value can be unintentionally generated.
+See the [Flax issue](https://github.com/google/flax/issues/2157) for more details.
+
+```{code-cell}
+:outputId: c11f51c1-1998-47f9-b041-7a611c73d169
+
+class Leaf(nn.Module):
+  def __call__(self, x):
+    return x + jax.random.randint(self.make_rng("rng"), (), 0, 100)
+
+class Node(nn.Module):
+  leaf_name: str
+  @nn.compact
+  def __call__(self, x):
+    return Leaf(name=self.leaf_name)(x)
+
+class Model(nn.Module):
+  @nn.compact
+  def __call__(self, x):
+    return (Node(name="ab", leaf_name="cdef")(x),
+            Node(name="abc", leaf_name="def")(x),
+    )
+
+out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})
+out1 == out2 # same output, despite having different submodule names
+```
+
+This occurs because the hash function [concatenates strings together](https://docs.python.org/3/library/hashlib.html#hashlib.hash.update), so the data `('AB', 'C')` is equivalent to data `('A', 'BC')` when fed into the hash function, therefore producing the same hash int.
+
+```{code-cell}
+:outputId: 1ed17f72-43cc-4f00-95af-5231f13a7f4b
+
+print(produce_hash(data=('A', 'B', 'C', 1)))
+print(produce_hash(data=('AB', 'C', 1)))
+print(produce_hash(data=('A', 'BC', 1)))
+print(produce_hash(data=('ABC', 1)))
+```
+
+To avoid this edge case, users can flip the `flax_fix_rng_separator` [configuration flag](https://flax.readthedocs.io/en/latest/api_reference/flax.config.html#flax.configurations.Config.flax_fix_rng_separator) to `True`.
+
+```{code-cell}
+:outputId: 21e623da-91d2-4bbc-e167-d3b257344124
+
+flax.config.update('flax_fix_rng_separator', True)
+out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})
+out1 == out2 # different output
+```
+
+## Managing RNG's on multiple devices
```

Comment:
Wondering if we should recommend `shard_map` instead?

Reply:
what about having both an example for `jit` and and for `shard_map`?

Reply:
+1 to both `(p)jit` and `shard_map`